Refactor using Scraped

ondenman opened the pull request:
Refactoring to use Scraped.

## [Scraper Change checklist](https://github.com/everypolitician/everypolitician/wiki/Scraper-Change-checklist)
* [x] 1. scraper is on Morph.io under the "everypolitician-scrapers" group?
* [x] 2. scraper's GitHub "Website" link points at morph.io page?
* [x] 3. scraper is set to auto-run?
* [x] 4. scraper is archiving?
* [x] 5. legislature has a scraper webhook set?

## [Gemfile checklist](https://github.com/everypolitician/everypolitician/wiki/Gemfile-checklist/)
* [x] 1. all links are secure
* [x] 2. links to Github use `github:` protocol, not simply `git:`
* [x] 3. formatting is consistent with our normal Rubocop setup

This is a step towards incorporating the scraping of the same data in Burmese which is available from the same source as a separate JSON set.

ondenman committed:
Tidy Gemfile for Rubocop

ondenman committed:
Add scraped gem

ondenman committed:
Bundle install

ondenman committed:
Add Scraped::JSON

The source site serves data as JSON. Subclassing from this class
will allow us to read that JSON without further parsing. Takes json as a keyword argument.

ondenman committed:
Extract MembersRecords from Members

Members are listed across multiple files. MembersRecords represents
the data in a single JSON file.

Where Members was returning a list of members from all the JSON files,
MembersRecords only returns the lower house members found in a single
JSON file. We now combine all the JSON, outside of the class, in scraper.rb.

ondenman committed:
Write to db

Iterates over all the members and writes their data to the db.

ondenman committed:
Rename Member to MemberRecords

The class represents the records of the member, not the
member themself. I have renamed the class to reflect this.

ondenman committed:
Refactor MemberRecords to use Scraped

ondenman committed:
Add String#tidy

ondenman committed:
Add scraped page archive

ondenman committed:
Tidy for Rubocop

tmtmtmtm commented on Gemfile:
```diff
@@ -6,12 +6,14 @@ git_source(:github) { |repo_name| "https://github.com/#{repo_name}.git" }
 ruby '2.3.1'
 
 gem 'colorize'
+gem 'combine_popolo_memberships', git: 'https://github.com/everypolitician/combine_popolo_memberships.git'
```
You're changing a Gemfile without including the Gemfile checklist… (which would in turn make it clearer what mistake you're making here…)

tmtmtmtm commented on scraper.rb:
```diff
@@ -3,12 +3,28 @@
+    r << current_page.members_of_the_lower_house
+  end
+  r.flatten
+end
```
This is all quite complex, but as it's not part of any public interface, I'm curious why you actually _need_ to build up the list of all members first, before writing them to the database, rather than simply iterating through all the pages and saving as you go.

tmtmtmtm reviewed (changes requested):
I'd quite like to hear from @chrismytton about the general approach here. My suspicion is that we might want to add a `Scraped::JSON` to `Scraped` itself, parallel to `Scraped::HTML`, but I'd not sure whether we want to do that there as part of this, or use this as a test case first, etc.

ondenman committed:
Scrape MembersList pages recursively

Instead of building up an array, we scrape each page
and write to the db as long as the current page has a :next field.

ondenman committed:
Remove superfluous requires

These Gems are no longer required by the scraper.

ondenman committed:
Remove superfluous gems from gemfile

ondenman committed:
Bundle install

tmtmtmtm commented on Gemfile:
```diff
@@ -5,13 +5,12 @@ git_source(:github) { |repo_name| "https://github.com/#{repo_name}.git" }
 
 ruby '2.3.1'
```
It's not _totally_ necessary, but when you're doing a refactoring of this scale, it's probably worth bringing the ruby version up to our current standard (i.e. 2.3.3) whilst you're at it.

tmtmtmtm commented on lib/Scraped/json.rb:
```diff
@@ -0,0 +1,18 @@
+require 'scraped'
+
+class Scraped
```
This is only working because you're running off an old version of `scraped`. `Scraped` is module, not a class.

tmtmtmtm commented on lib/member_records.rb:
```diff
@@ -1,20 +1,14 @@
   def tidy
     gsub(/[[:space:]]+/, ' ').strip
   end
 end
```
If you're using a modern `Scraped` (which you should be) you can remove this.

tmtmtmtm commented on lib/members_records.rb:
```diff
@@ -2,45 +2,30 @@
-      Member.new(mem)
+      MemberRecords.new(response: response, json: mem)
     end
   end
```
This looks like you want a 'fragment' method.

tmtmtmtm reviewed (changes requested):
I'm finding the naming conventions very difficult here, particularly the pluralisation on 'MemberRecords', when this seems to be dealing with a single record, not more than one.

In both cases, there also seems to be a bit of a mismatch in that the data you're dealing with is records of *Memberships*, not Members.

ondenman committed:
Bump ruby version

ondenman committed:
Bundle update scraped

ondenman committed:
Subclass Scraped::JSON from Scraped::Document

We're now using the latest version of Scraped. Scraped is
now a module. We want Scraped::JSON to inherit from Scraped::Document.

ondenman committed:
Do not require field_serializer in MembersRecords

The gem is included in Scraped.

ondenman committed:
Add Scraped::JSON#fragment

So that we can create fragments of a MembersList page.

ondenman committed:
Create MemberRecords as a JSON fragment

ondenman committed:
Rename member_records to membership_record

The class represents a single membership record so the name
has been changed to represent this.

ondenman committed:
Rename MembersRecords to MembershipRecords

The class represents a collection of membership records. The name
has been changed to reflect this.

ondenman committed:
Clear out database before scraping

tmtmtmtm commented on lib/Scraped/json.rb:
```diff
@@ -0,0 +1,23 @@
+    end
+
+    def json
+      @json ||= ::JSON.parse(Nokogiri::HTML(response.body), symbolize_names: true)
```
Why are you going through `Nokogiri::HTML` here, when this _isn't_ HTML?

tmtmtmtm commented on lib/Scraped/json.rb:
```diff
@@ -0,0 +1,23 @@
+    end
+
+    def fragment(mapping)
+      noko_fragment, klass = mapping.to_a.first
```
`noko_fragment` is a rather odd name here, when this is json, and nothing to do with nokogiri. If this is to deliberately highlight that it's lifted from elsewhere, then I suspect you're better either removing the need for the duplication, though that might be best happening upstream (as it'll probably involve making this and `Scraped::HTML` have a common base class), or leaving a comment behind to explain to later readers what's going on.

tmtmtmtm commented on lib/membership_records.rb:
```diff
@@ -0,0 +1,30 @@
+    json[:next]
+  end
+
+  def json_for(url)
```
What is this method? It seems quite out of place here, but I can't see where it's actually used.

tmtmtmtm commented on scraper.rb:
```diff
@@ -1,14 +1,22 @@
 
-members.each do |member|
-  ScraperWiki.save_sqlite([:id], member.to_h)
+def all_members(url)
```
This is oddly named. It seems like it should be a query method, but this is actually a command method, with no return value, whose job is to save the data.

ondenman committed:
Parse JSON document body without going through Nokogiri

There's no need to go through Nokogiri as the body is JSON.

ondenman committed:
Rename noko_fragment as json_fragment

This is a more appropriate name as we are dealing with JSON, not
a nokogiri document.

ondenman committed:
Remove superfluous method

Method is no longer used

ondenman committed:
Rename all_members(url) to scrape_all_members(url)

The new name reflects that the method is a command.

tmtmtmtm commented on lib/membership_records.rb:
```diff
@@ -0,0 +1,26 @@
+require 'json'
+require_relative 'membership_record'
+
+OpenURI::Cache.cache_path = '.cache'
```
Why do you need all this OpenURI related setup here?

tmtmtmtm commented on scraper.rb:
```diff
@@ -1,14 +1,21 @@
 
-members.each do |member|
-  ScraperWiki.save_sqlite([:id], member.to_h)
+def scrape_all_members(url)
```
This is confusingly named. In conjunction with similarly named 'current_members", and "members_of_the_lower_house", the implication is that this is dealing with _all_ members. And thus it's confusing when it calls itself repeatedly. It would be good to make it more obvious that this is about pagination.

tmtmtmtm committed:
Initial rubocop configuration

tmtmtmtm committed:
Add a 'rake rubocop' and configure travis to use it

tmtmtmtm reviewed (changes requested):
NB: this code doesn't pass rubocop checking. I've made travis check for that automatically now.

ondenman committed:
Remove superfluous OpenURI setup

ondenman committed:
Use less confusing names

New names make it clear that we're scraping members of the lower house
page by page.

ondenman committed:
fixup! Add scraped page archive

ondenman committed:
Tidy for Rubocop

ondenman committed:
Add rake gem

ondenman committed:
Bundle install

ondenman commented:
Closing in favour of: https://github.com/everypolitician-scrapers/myanmar-lower-house-2016/pull/8#event-972411920
